src/hotspot/share/prims/jvmtiEnv.cpp

```diff
@@ -1634,107 +1634,20 @@ JvmtiEnv::GetFrameCount(JavaThread* java_thread, jint* count_ptr) {
-    // are Java (not native) frames, and that there is no intervening VM frame
-    int frame_count = 0;
-    bool is_interpreted[2];
-    intptr_t *frame_sp[2];
-    // The 2-nd arg of constructor is needed to stop iterating at java entry frame.
-    for (vframeStream vfs(java_thread, true, false /* process_frames */); !vfs.at_end(); vfs.next()) {
-      methodHandle mh(current_thread, vfs.method());
-      if (mh->is_native()) return(JVMTI_ERROR_OPAQUE_FRAME);
-      is_interpreted[frame_count] = vfs.is_interpreted_frame();
-      frame_sp[frame_count] = vfs.frame_id();
-      if (++frame_count > 1) break;
-    }
-    if (frame_count < 2)  {
-      // We haven't found two adjacent non-native Java frames on the top.
-      // There can be two situations here:
-      //  1. There are no more java frames
-      //  2. Two top java frames are separated by non-java native frames
-      if(vframeForNoProcess(java_thread, 1) == NULL) {
-        return JVMTI_ERROR_NO_MORE_FRAMES;
-      } else {
-        // Intervening non-java native or VM frames separate java frames.
-        // Current implementation does not support this. See bug #5031735.
-        // In theory it is possible to pop frames in such cases.
-        return JVMTI_ERROR_OPAQUE_FRAME;
-      }
-    }
-
-    // If any of the top 2 frames is a compiled one, need to deoptimize it
-    for (int i = 0; i < 2; i++) {
-      if (!is_interpreted[i]) {
-        Deoptimization::deoptimize_frame(java_thread, frame_sp[i]);
-      }
-    }
-
-    // Update the thread state to reflect that the top frame is popped
-    // so that cur_stack_depth is maintained properly and all frameIDs
-    // are invalidated.
-    // The current frame will be popped later when the suspended thread
-    // is resumed and right before returning from VM to Java.
-    // (see call_VM_base() in assembler_<cpu>.cpp).
-
-    // It's fine to update the thread state here because no JVMTI events
-    // shall be posted for this PopFrame.
-
-    // It is only safe to perform the direct operation on the current
-    // thread. All other usage needs to use a handshake for safety.
-    {
-      MutexLocker mu(JvmtiThreadState_lock);
-      if (java_thread == JavaThread::current()) {
-        state->update_for_pop_top_frame();
-      } else {
-        UpdateForPopTopFrameClosure op(state);
-        Handshake::execute(&op, java_thread);
-        if (op.result() != JVMTI_ERROR_NONE) {
-          return op.result();
-        }
-      }
-    }
-
-    java_thread->set_popframe_condition(JavaThread::popframe_pending_bit);
-    // Set pending step flag for this popframe and it is cleared when next
-    // step event is posted.
-    state->set_pending_step_for_popframe();
+  MutexLocker mu(JvmtiThreadState_lock);
+  UpdateForPopTopFrameClosure op(state);
+  if (java_thread == JavaThread::current()) {
+    op.doit(java_thread, true);
+  } else {
+    Handshake::execute(&op, java_thread);
```

Comment:
This pattern is repeated a lot - we should be able to incorporate it into the op itself by passing in `java_thread`.

Reply:
My suggestion here is that we fix this in Handshake::execute() in separate RFE.